src/main.rs

```diff
@@ -69,29 +73,62 @@ impl Client {
     // Method should return an error when a connection already exists.
     // The client should send a handshake to the server.
     fn open(&mut self, addr: &str, server: Server) -> CommsResult<()> {
-        todo!()
+        if self.connections.contains_key(addr) {
+            return Err(CommsError::ConnectionExists(addr.to_string()));
+        }
+
+        let msg = Message {
+            msg_type: MessageType::Handshake,
+            load: self.ip.clone(),
+        };
+
+        let mut svr = server;
+
+        match svr.receive(msg) {
```

Comment:
Metoda `open()` nie powinna posiadać w sobie logiki porozumiewania się z serwerem. Do tego służy już inna metoda, `send()`, i to z niej powinno korzystać `open()`.